### ISSUE TITLE
ipfs-http-client doesn't work in the browser with Meteor 

### ISSUE THREAD

**Issue by yurtsiv** [ipfs/js-ipfs#3557]
- **Version**: `49.0.2`

- **Platform**: 
`Linux hostname  5.10.15-1-MANJARO #1 SMP PREEMPT Wed Feb 10 10:42:47 UTC 2021 x86_64 GNU/Linux`

#### Description:

I’m having troubles using the latest version of `ipfs-http-client` with Meteor on the client.
It fails with: `Object prototype may only be an Object or null: undefined`.
This error comes from `graceful-fs` package. The dependency flow is as following:

`ipfs-http-client` -> `ipfs-utils` -> `fs-extra` -> `graceful-fs`

The lates working version is `41.0.1`.

#### Steps to reproduce the error:

I've created a [minimal reproduction](https://github.com/yurtsiv/meteor-ipfs-problem) of the issue. All the steps to reproduce the error are provided in README. Let me know if you need more details.

Thanks!

**Comment by achingbrain** [comment]
That's interesting - it's probably trying to pull node-specific code into a browser bundle.

Can you configure Meteor to follow browser semantics for building the front end code?

**Comment by yurtsiv** [comment]
@achingbrain Yes, that's the case.

The [meteor build system](https://guide.meteor.com/build-tool.html) is not as configurable as say webpack. They provide [meteor-node-stubs](https://github.com/meteor/node-stubs) which mocks some of the Node.js modules, but the `fs` [is set to null](https://github.com/meteor/node-stubs/blob/master/map.json#L13).

I'm wondering, how does it work with webpack. Does the code which contains references to `fs` gets shipped to the browser with some stub implementation of `fs` or it just gets stripped out?

**Comment by achingbrain** [comment]
`fs-extra` is used by `ipfs-utils` in the `src/files/glob-source.js` file.  Glob sources are only used when importing from a filesystem so shouldn't be used in the browser.

Webpack respects the `browser` field in `package.json` which for `ipfs-http-client` [sets src/files/glob-source.js to `false`](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs-http-client/package.json#L19) so it doesn't get pulled in to the bundle which would pull `fs-extra` in with it.

**Comment by yurtsiv** [comment]
Ok, turns out Meteor does respect `browser` field but only for relative paths (https://github.com/meteor/meteor/issues/6890#issuecomment-339817703, https://github.com/meteor/meteor/pull/9311).

Wouldn't it be better to specify browser specific files for a particular package in the package itself, thus reducing coupling? In this case move [the declaration](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs-http-client/package.json#L19) to `ipfs-utils` itself.

**Comment by achingbrain** [comment]
> Wouldn't it be better to specify browser specific files for a particular package in the package itself, thus reducing coupling?

Yes - could you please open a PR to the affected modules to make this change?

**Comment by yurtsiv** [comment]
Sure. Will do this in a few days.

**Comment by yurtsiv** [comment]
Maybe a few days more :sweat_smile: Really busy currently

**Comment by yurtsiv** [comment]
@achingbrain Just wondering, why [js-ipfs-utils](https://github.com/ipfs/js-ipfs-utils) is not a part of this repo?

**Comment by achingbrain** [comment]
It was briefly but it’s used in too many places so needed it’s own release schedule.

**Comment by yurtsiv** [comment]
Hi. Can you take a look at the following PRs, please: https://github.com/ipfs/js-ipfs/pull/3630, https://github.com/ipfs/js-ipfs-utils/pull/121

**Comment by yurtsiv** [comment]
To continue the discussion on https://github.com/ipfs/js-ipfs/pull/3630, I don't have a solution for `normalise-input`, BUT I do think it makes sense to at least get rid of

```
{
   ...
   "browser": {
      ...
      "ipfs-utils/src/files/glob-source": false
      ...
    }
}
```

in every package under `js-ipfs/packages`, since it's already [handled](https://github.com/ipfs/js-ipfs-utils/blob/master/package.json#L18) in `ipfs-utils` (https://github.com/ipfs/js-ipfs-utils/pull/121). No need to have it twice.

Please confirm that my reasoning is correct and I'll make a PR.

**Comment by whizzzkid** [comment]
js-ipfs is being deprecated in favor of [Helia](https://github.com/ipfs/helia). You can https://github.com/ipfs/js-ipfs/issues/4336 and [read the migration guide](https://github.com/ipfs/helia/wiki/%F0%9F%9A%9B-Migrating-from-js-IPFS).

Please feel to reopen with any comments before 2023-06-05. We will do a final pass on reopened issues afterward (see https://github.com/ipfs/js-ipfs/issues/4336).
